main/src/ca/uwaterloo/flix/language/phase/optimizer/Inliner.scala

```diff
@@ -497,8 +501,11 @@ object Inliner {
     * @param exps the arguments to the function.
     * @param ctx0 the local context.
     */
-  private def shouldInlineDef(defn: MonoAst.Def, exps: List[Expr], ctx0: LocalContext): Boolean = {
-    !ctx0.currentlyInlining && !defn.spec.defContext.isSelfRef &&
+  private def shouldInlineDef(defn: MonoAst.Def, exps: List[Expr], ctx0: LocalContext)(implicit sym0: Symbol.DefnSym, inlined: ConcurrentHashMap[Symbol.DefnSym, ConcurrentHashMap[Symbol.DefnSym, Unit]]): Boolean = {
+    val isInlinedHere = inlined.getOrDefault(defn.sym, new ConcurrentHashMap()).containsKey(sym0)
+    !defn.spec.ann.isDontInline &&
+      ((defn.spec.ann.isInline && defn.spec.defContext.isSelfRef && !isInlinedHere) || !defn.spec.defContext.isSelfRef || (defn.spec.ann.isInline && !defn.spec.defContext.isSelfRef)) &&
+      !ctx0.currentlyInlining &&
```

Comment:
This is btw not atomic so a race condition could inline `defn` multiple places

Reply:
Why do we need the map?

Reply:
We need to know if it has been unrolled once already in `sym0`

Reply:
The new solution is somewhat verbose but that's the nature of java and null checks I guess...